File: app/src/main/java/com/battlelancer/seriesguide/ui/search/TraktAddFragment.java
package com.battlelancer.seriesguide.ui.search;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.ui.OverviewActivity;
import com.battlelancer.seriesguide.ui.SearchActivity;
import com.battlelancer.seriesguide.ui.shows.ShowTools;
import com.battlelancer.seriesguide.util.TaskManager;
import com.battlelancer.seriesguide.util.tasks.BaseShowActionTask;
import com.battlelancer.seriesguide.widgets.EmptyView;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

/**
 * Can display either the connected trakt user's watched, collected or watchlist-ed shows and offer
 * to add them.
 */
public class TraktAddFragment extends AddFragment {

    /**
     * Which trakt list should be shown. One of {@link TraktShowsLink}.
     */
    private final static String ARG_TYPE = "traktListType";

    public static TraktAddFragment newInstance(TraktShowsLink link) {
        TraktAddFragment f = new TraktAddFragment();

        Bundle args = new Bundle();
        args.putInt(TraktAddFragment.ARG_TYPE, link.id);
        f.setArguments(args);

        return f;
    }

    private Unbinder unbinder;
    private TraktShowsLink listType;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        listType = TraktShowsLink.fromId(args != null ? args.getInt(ARG_TYPE) : -1);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_addshow_trakt, container, false);
        unbinder = ButterKnife.bind(this, v);

        // set initial view states
        setProgressVisible(true, false);

        return v;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        // setup adapter, enable context menu only for watchlist
        adapter = new AddAdapter(getActivity(), new ArrayList<>(), itemClickListener,
                listType == TraktShowsLink.WATCHLIST
        );

        // load data
        LoaderManager.getInstance(this)
                .initLoader(SearchActivity.TRAKT_BASE_LOADER_ID + listType.id, null,
                        traktAddCallbacks);

        // add menu options
        setHasOptionsMenu(true);
    }

    private final AddAdapter.OnItemClickListener itemClickListener
            = new AddAdapter.OnItemClickListener() {

        @Override
        public void onItemClick(SearchResult item) {
            if (item != null && item.getState() != SearchResult.STATE_ADDING) {
                if (item.getState() == SearchResult.STATE_ADDED) {
                    // already in library, open it
                    startActivity(OverviewActivity
                            .intentShowByTmdbId(requireContext(), item.getTmdbId()));
                } else {
                    // display more details in a dialog
                    AddShowDialogFragment.show(getParentFragmentManager(), item);
                }
            }
        }

        @Override
        public void onAddClick(SearchResult item) {
            EventBus.getDefault().post(new OnAddingShowEvent(item.getTmdbId()));
            TaskManager.getInstance().performAddTask(requireContext(), item);
        }

        @Override
        public void onMenuWatchlistClick(View view, int showTmdbId) {
            PopupMenu popupMenu = new PopupMenu(view.getContext(), view);
            popupMenu.inflate(R.menu.add_dialog_popup_menu);

            // prevent adding shows to watchlist already on watchlist
            if (listType == TraktShowsLink.WATCHLIST) {
                popupMenu.getMenu().findItem(R.id.menu_action_show_watchlist_add).setVisible(false);
            }

            popupMenu.setOnMenuItemClickListener(
                    new AddItemMenuItemClickListener(requireContext(), showTmdbId));
            popupMenu.show();
        }
    };

    public static class AddItemMenuItemClickListener implements PopupMenu.OnMenuItemClickListener {

        private final Context context;
        private final int showTmdbId;

        public AddItemMenuItemClickListener(Context context, int showTmdbId) {
            this.context = context;
            this.showTmdbId = showTmdbId;
        }

        @Override
        public boolean onMenuItemClick(MenuItem item) {
            int itemId = item.getItemId();
            if (itemId == R.id.menu_action_show_watchlist_add) {
                ShowTools.addToWatchlist(context, showTmdbId);
                return true;
            }
            if (itemId == R.id.menu_action_show_watchlist_remove) {
                ShowTools.removeFromWatchlist(context, showTmdbId);
                return true;
            }
            return false;
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.trakt_library_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.menu_add_all) {
            if (searchResults != null) {
                List<SearchResult> showsToAdd = new LinkedList<>();
                // only include shows not already added
                for (SearchResult result : searchResults) {
                    if (result.getState() == SearchResult.STATE_ADD) {
                        showsToAdd.add(result);
                        result.setState(SearchResult.STATE_ADDING);
                    }
                }
                EventBus.getDefault().post(new OnAddingShowEvent());
                TaskManager.getInstance().performAddTask(getContext(), showsToAdd, false, false);
            }
            // disable the item so the user has to come back
            item.setEnabled(false);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        unbinder.unbind();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(BaseShowActionTask.ShowChangedEvent event) {
        if (listType == TraktShowsLink.WATCHLIST) {
            // reload watchlist if a show was removed
            LoaderManager.getInstance(this)
                    .restartLoader(SearchActivity.TRAKT_BASE_LOADER_ID + listType.id, null,
                            traktAddCallbacks);
        }
    }

    @Override
    protected void setupEmptyView(EmptyView emptyView) {
        emptyView.setButtonClickListener(v -> {
            setProgressVisible(true, false);
            LoaderManager.getInstance(TraktAddFragment.this)
                    .restartLoader(SearchActivity.TRAKT_BASE_LOADER_ID + listType.id, null,
                            traktAddCallbacks);
        });
    }

    private final LoaderManager.LoaderCallbacks<TraktAddLoader.Result> traktAddCallbacks
            = new LoaderManager.LoaderCallbacks<TraktAddLoader.Result>() {
        @Override
        public Loader<TraktAddLoader.Result> onCreateLoader(int id, Bundle args) {
            return new TraktAddLoader(getContext(), listType);
        }

        @Override
        public void onLoadFinished(@NonNull Loader<TraktAddLoader.Result> loader,
                TraktAddLoader.Result data) {
            if (!isAdded()) {
                return;
            }
            setSearchResults(data.results);
            setEmptyMessage(data.emptyText);
            setProgressVisible(false, true);
        }

        @Override
        public void onLoaderReset(@NonNull Loader<TraktAddLoader.Result> loader) {
            // keep currently displayed data
        }
    };
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/search/discover/TraktAddFragment.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2011-2024 Uwe Trottmann

package com.battlelancer.seriesguide.shows.search.discover

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ArrayAdapter
import android.widget.GridView
import androidx.core.view.MenuProvider
import androidx.core.view.ViewCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.loader.app.LoaderManager
import androidx.loader.content.Loader
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.FragmentAddshowTraktBinding
import com.battlelancer.seriesguide.enums.NetworkResult
import com.battlelancer.seriesguide.shows.tools.AddShowTask
import com.battlelancer.seriesguide.shows.tools.ShowTools2
import com.battlelancer.seriesguide.ui.widgets.EmptyView
import com.battlelancer.seriesguide.util.TaskManager
import com.battlelancer.seriesguide.util.ThemeUtils
import com.battlelancer.seriesguide.util.tasks.BaseShowActionTask.ShowChangedEvent
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.LinkedList

/**
 * Displays the watched shows, show collection or shows watchlist of the connected Trakt user for
 * the purpose of adding them. Shows on the watchlist can also be removed from the watchlist.
 */
class TraktAddFragment : Fragment() {

    private var binding: FragmentAddshowTraktBinding? = null
    private val contentContainer: View
        get() = binding!!.containerAddContent

    private val progressBar: View
        get() = binding!!.progressBarAdd

    private val resultsGridView: GridView
        get() = binding!!.gridViewAdd

    private val emptyView: EmptyView
        get() = binding!!.emptyViewAdd

    private lateinit var listType: TraktAddLoader.Type
    private var searchResults: List<SearchResult>? = null
    private lateinit var adapter: AddAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val typeOrdinal = requireArguments().getInt(ARG_TYPE)
        listType = TraktAddLoader.Type.entries.find { it.ordinal == typeOrdinal }
            ?: throw IllegalArgumentException("Invalid Type ordinal $typeOrdinal")
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = FragmentAddshowTraktBinding.inflate(inflater, container, false)
        .also { binding = it }
        .root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        emptyView.setButtonClickListener {
            setProgressVisible(visible = true, animate = false)
            LoaderManager.getInstance(this@TraktAddFragment)
                .restartLoader(
                    ShowsTraktActivity.TRAKT_BASE_LOADER_ID + listType.ordinal, null,
                    traktAddCallbacks
                )
        }

        // basic setup of grid view
        resultsGridView.emptyView = emptyView
        // enable app bar scrolling out of view
        ViewCompat.setNestedScrollingEnabled(resultsGridView, true)

        ThemeUtils.applyBottomPaddingForNavigationBar(resultsGridView)
        ThemeUtils.applyBottomMarginForNavigationBar(binding!!.textViewPoweredByAddShowTrakt)

        // set initial view states
        setProgressVisible(visible = true, animate = false)

        // set up adapter
        adapter = AddAdapter(
            requireContext(), ArrayList(), itemClickListener,
            listType == TraktAddLoader.Type.WATCHLIST
        )

        // load data
        LoaderManager.getInstance(this)
            .initLoader(
                ShowsTraktActivity.TRAKT_BASE_LOADER_ID + listType.ordinal, null,
                traktAddCallbacks
            )

        // add menu options
        requireActivity().addMenuProvider(
            optionsMenuProvider,
            viewLifecycleOwner,
            Lifecycle.State.RESUMED
        )
    }

    /**
     * Hides the content container and shows a progress bar.
     */
    private fun setProgressVisible(visible: Boolean, animate: Boolean) {
        if (animate) {
            val fadeOut = AnimationUtils.loadAnimation(activity, android.R.anim.fade_out)
            val fadeIn = AnimationUtils.loadAnimation(activity, android.R.anim.fade_in)
            contentContainer.startAnimation(if (visible) fadeOut else fadeIn)
            progressBar.startAnimation(if (visible) fadeIn else fadeOut)
        }
        contentContainer.visibility = if (visible) View.GONE else View.VISIBLE
        progressBar.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    private val itemClickListener
        get() = object :
            ItemAddShowClickListener(requireContext(), lifecycle, parentFragmentManager) {

            override fun onMoreOptionsClick(view: View, show: SearchResult) {
                AddShowPopupMenu(requireContext(), show, view).apply {
                    // For watchlist only show remove from, for other lists show no watchlist option
                    hideAddToWatchlistAction()
                    if (listType != TraktAddLoader.Type.WATCHLIST) {
                        hideRemoveFromWatchlistAction()
                    }
                }.show()
            }
        }

    private val optionsMenuProvider: MenuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.trakt_library_menu, menu)
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            val itemId = menuItem.itemId
            if (itemId == R.id.menu_add_all) {
                val searchResults = searchResults
                if (searchResults != null) {
                    val showsToAdd = LinkedList<AddShowTask.Show>()
                    // only include shows not already added
                    for (result in searchResults) {
                        if (result.state == SearchResult.STATE_ADD) {
                            showsToAdd.add(
                                AddShowTask.Show(
                                    result.tmdbId,
                                    result.languageCode,
                                    result.title
                                )
                            )
                            result.state = SearchResult.STATE_ADDING
                        }
                    }
                    EventBus.getDefault().post(OnAddingShowEvent())
                    TaskManager.performAddTask(
                        requireContext(), showsToAdd,
                        isSilentMode = false,
                        isMergingShows = false
                    )
                }
                // disable the item so the user has to come back
                menuItem.isEnabled = false
                return true
            }
            return false
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(@Suppress("UNUSED_PARAMETER") event: ShowChangedEvent?) {
        if (listType == TraktAddLoader.Type.WATCHLIST) {
            // reload watchlist if a show was removed
            LoaderManager.getInstance(this)
                .restartLoader(
                    ShowsTraktActivity.TRAKT_BASE_LOADER_ID + listType.ordinal, null,
                    traktAddCallbacks
                )
        }
    }

    /**
     * Called if the user triggers adding a single new show through the add dialog. The show is not
     * actually added, yet.
     *
     * See [onEvent(OnShowAddedEvent)][onEvent].
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: OnAddingShowEvent) {
        if (event.showTmdbId > 0) {
            adapter.setStateForTmdbId(event.showTmdbId, SearchResult.STATE_ADDING)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: AddShowTask.OnShowAddedEvent) {
        if (event.successful) {
            setShowAdded(event.showTmdbId)
        } else if (event.showTmdbId > 0) {
            setShowNotAdded(event.showTmdbId)
        } else {
            adapter.setAllPendingNotAdded()
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: ShowTools2.OnShowRemovedEvent) {
        if (event.resultCode == NetworkResult.SUCCESS) {
            setShowNotAdded(event.showTmdbId)
        }
    }

    private fun setShowAdded(showTmdbId: Int) {
        adapter.setStateForTmdbId(showTmdbId, SearchResult.STATE_ADDED)
    }

    private fun setShowNotAdded(showTmdbId: Int) {
        adapter.setStateForTmdbId(showTmdbId, SearchResult.STATE_ADD)
    }

    private val traktAddCallbacks: LoaderManager.LoaderCallbacks<TraktAddLoader.Result> =
        object : LoaderManager.LoaderCallbacks<TraktAddLoader.Result> {
            override fun onCreateLoader(id: Int, args: Bundle?): Loader<TraktAddLoader.Result> {
                return TraktAddLoader(requireContext(), listType)
            }

            override fun onLoadFinished(
                loader: Loader<TraktAddLoader.Result>,
                data: TraktAddLoader.Result
            ) {
                updateSearchResults(data.results, data.emptyText)
                setProgressVisible(visible = false, animate = true)
            }

            override fun onLoaderReset(loader: Loader<TraktAddLoader.Result>) {
                // keep currently displayed data
            }
        }

    private fun updateSearchResults(results: List<SearchResult>, emptyText: CharSequence) {
        this.searchResults = results
        adapter.clear()
        adapter.addAll(results)
        resultsGridView.adapter = adapter
        emptyView.setMessage(emptyText)
    }

    companion object {
        /**
         * Which Trakt list should be shown. Ordinal of one of [TraktAddLoader.Type].
         */
        private const val ARG_TYPE = "traktListType"

        val liftOnScrollTargetViewId = R.id.gridViewAdd

        fun newInstance(type: TraktAddLoader.Type): TraktAddFragment {
            val f = TraktAddFragment()
            val args = Bundle()
            args.putInt(ARG_TYPE, type.ordinal)
            f.arguments = args
            return f
        }
    }

    class AddAdapter(
        context: Context,
        objects: List<SearchResult>,
        private val itemClickListener: ItemAddShowViewHolder.ClickListener,
        private val showWatchlistActions: Boolean
    ) : ArrayAdapter<SearchResult>(context, 0, objects) {

        private fun getItemForShowTmdbId(showTmdbId: Int): SearchResult? {
            val count = count
            for (i in 0 until count) {
                val item = getItem(i)
                if (item != null && item.tmdbId == showTmdbId) {
                    return item
                }
            }
            return null
        }

        fun setStateForTmdbId(showTmdbId: Int, state: Int) {
            val item = getItemForShowTmdbId(showTmdbId)
            if (item != null) {
                item.state = state
                notifyDataSetChanged()
            }
        }

        fun setAllPendingNotAdded() {
            val count = count
            for (i in 0 until count) {
                val item = getItem(i)
                if (item != null && item.state == SearchResult.STATE_ADDING) {
                    item.state = SearchResult.STATE_ADD
                }
            }
            notifyDataSetChanged()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view: View
            // Re-use ItemAddShowViewHolder, but not using the RecyclerView features
            val holder: ItemAddShowViewHolder
            if (convertView == null) {
                holder = ItemAddShowViewHolder.create(parent, itemClickListener)
                    .also { it.itemView.tag = it }
                view = holder.itemView
            } else {
                holder = convertView.tag as ItemAddShowViewHolder
                view = convertView
            }

            val item = getItem(position)
            holder.bindTo(context, item, showWatchlistActions)

            return view
        }

    }
}